File: tensorflow/core/ops/risc_ops.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/shape_inference.h"

namespace tensorflow {

namespace {
Status RiscBinaryNonBroadcastOpShapeFn(shape_inference::InferenceContext* c) {
  const auto rank = c->Rank(c->input(0));
  if (rank != c->Rank(c->input(1))) {
    return errors::InvalidArgument("Mismatch rank for input.");
  }
  for (int i = 0; i < rank; ++i) {
    if (!c->ValueKnown(c->Dim(c->input(0), i)) ||
        !c->ValueKnown(c->Dim(c->input(1), i))) {
      continue;
    }
    if (c->Value(c->Dim(c->input(0), i)) != c->Value(c->Dim(c->input(1), i))) {
      return errors::InvalidArgument("Mismatch shapes for input.");
    }
  }
  c->set_output(0, c->input(0));
  auto* handle_data = c->input_handle_shapes_and_types(0);
  if (handle_data != nullptr) {
    c->set_output_handle_shapes_and_types(0, *handle_data);
  }
  return Status::OK();
}
}  // namespace

REGISTER_OP("RiscAbs")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscAdd")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn)
    .SetIsAggregate()
    .SetIsCommutative();

// TODO(b/178234771): retire this.
REGISTER_OP("RiscBinaryArithmetic")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("op_type: {'ADD', 'SUB', 'MUL', 'DIV', 'REM', 'MIN', 'POW'}")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscBinaryComparison")
    .Input("x: T")
    .Input("y: T")
    .Output("z: bool")
    .Attr("op_type: {'EQ', 'NE', 'GE', 'GT', 'LE', 'LT'}")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscBitcast")
    .Input("x: SrcT")
    .Output("y: DstT")
    .Attr("SrcT: type")
    .Attr("DstT: type")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscBroadcast")
    .Input("input: T")
    .Input("shape: Tidx")
    .Output("output: T")
    .Attr("T: type")
    .Attr("Tidx: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscCast")
    .Input("x: SrcT")
    .Output("y: DstT")
    .Attr("SrcT: type")
    .Attr("DstT: type")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscCeil")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscCholesky")
    .Input("input: T")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscConcat")
    .Input("values: N * T")
    .Input("axis: Tidx")
    .Output("output: T")
    .Attr("N: int >= 2")
    .Attr("T: type")
    .Attr("Tidx: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::ConcatV2Shape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscCondition")
    .Input("pred: bool")
    .Input("input_true: SrcT")
    .Input("input_false: SrcT")
    .Output("output: DstT")
    .Attr("func_true: func")
    .Attr("func_false: func")
    .Attr("SrcT: {bfloat16, half, float, double}")
    .Attr("DstT: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscConv")
    .Input("input: T")
    .Input("filter: T")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("strides: list(int)")
    .Attr(GetConvnetDataFormatAttrString())
    .SetShapeFn(shape_inference::UnknownShape)
    .Attr("dilations: list(int) = [1, 1, 1, 1]");

REGISTER_OP("RiscCos")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscDiv")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

REGISTER_OP("RiscDot")
    .Input("a: T")
    .Input("b: T")
    .Output("product: T")
    .Attr("transpose_a: bool = false")
    .Attr("transpose_b: bool = false")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::MatMulShape);

REGISTER_OP("RiscExp")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscFft")
    .Input("input: Tcomplex")
    .Output("output: Tcomplex")
    .Attr("Tcomplex: {complex64, complex128} = DT_COMPLEX64")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscFloor")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscGather")
    .Input("params: Tparams")
    .Input("indices: Tindices")
    .Input("axis: Taxis")
    .Attr("batch_dims: int = 0")
    .Output("output: Tparams")
    .Attr("Tparams: type")
    .Attr("Tindices: {int32,int64}")
    .Attr("Taxis: {int32,int64}")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscImag")
    .Input("input: T")
    .Output("output: Tout")
    .Attr("T: {complex64, complex128} = DT_COMPLEX64")
    .Attr("Tout: {float, double} = DT_FLOAT")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscIsFinite")
    .Input("x: T")
    .Output("y: bool")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscLog")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscLogicalAnd")
    .Input("x: bool")
    .Input("y: bool")
    .Output("z: bool")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscLogicalNot")
    .Input("x: bool")
    .Output("z: bool")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscLogicalOr")
    .Input("x: bool")
    .Input("y: bool")
    .Output("z: bool")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscMax")
    .Input("x: T")
    .Input("y: T")
    .Output("max: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

REGISTER_OP("RiscMin")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

REGISTER_OP("RiscMul")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

REGISTER_OP("RiscNeg")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscPad")
    .Input("input: T")
    .Input("paddings: Tpaddings")
    .Input("constant_values: T")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("Tpaddings: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscPool")
    .Input("value: T")
    .Output("output: T")
    .Attr("ksize: list(int) >= 4")
    .Attr("strides: list(int) >= 4")
    .Attr("pooling_type: {'AVG', 'MAX'}")
    .Attr(GetConvnetDataFormatAttrString())
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscPow")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

REGISTER_OP("RiscRandomUniform")
    .Input("shape: T")
    .Output("output: float")
    .Attr("seed: int = 0")
    .Attr("T: {int32, int64}")
    .SetShapeFn(shape_inference::RandomShape);

REGISTER_OP("RiscReal")
    .Input("input: T")
    .Output("output: Tout")
    .Attr("T: {complex64, complex128} = DT_COMPLEX64")
    .Attr("Tout: {float, double} = DT_FLOAT")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscReduce")
    .Input("tensor: T")
    .Input("axis: Index")
    .Output("output: T")
    .Attr("reduce_type: {'MEAN', 'SUM'}")
    .Attr("Index: {int32,int64} = DT_INT32")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscRem")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscReshape")
    .Input("tensor: T")
    .Input("shape: Tshape")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("Tshape: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscReverse")
    .Input("tensor: T")
    .Input("axis: Tidx")
    .Output("output: T")
    .Attr("Tidx: {int32, int64} = DT_INT32")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscScatter")
    .Input("indices: Tindices")
    .Input("updates: T")
    .Input("shape: Tindices")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("Tindices: {int32, int64}")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscShape")
    .Input("input: T")
    .Output("output: out_type")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("out_type: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscSign")
    .Input("x: T")
    .Output("y: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

REGISTER_OP("RiscSlice")
    .Input("input: T")
    .Input("begin: Index")
    .Input("size: Index")
    .Output("output: T")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("Index: {int32,int64}")
    .SetShapeFn(shape_inference::SliceShape);

REGISTER_OP("RiscSort")
    .Input("input: T")
    .Input("axis: Index")
    .Output("output: T")
    .Attr("Index: {int32,int64} = DT_INT32")
    .Attr("T: {bfloat16, half, float, double}")
    .Attr("direction: {'ASCENDING', 'DESCENDING'}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscSqueeze")
    .Input("input: T")
    .Output("output: T")
    .Attr("T: type")
    .Attr("squeeze_dims: list(int) >= 0 = []")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("RiscSub")
    .Input("x: T")
    .Input("y: T")
    .Output("z: T")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(RiscBinaryNonBroadcastOpShapeFn);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscTranspose")
    .Input("x: T")
    .Input("perm: Tperm")
    .Output("y: T")
    .Attr("T: type")
    .Attr("Tperm: {int32, int64} = DT_INT32")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscTriangularSolve")
    .Input("matrix: T")
    .Input("rhs: T")
    .Output("output: T")
    .Attr("lower: bool = True")
    .Attr("adjoint: bool = False")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnknownShape);

// TODO(b/178234771): retire this.
REGISTER_OP("RiscUnary")
    .Input("x: T")
    .Output("y: T")
    .Attr(
        "op_type: {'ABL', 'CEIL', 'COS', 'EXP', 'FLOOR', 'IMAG', 'LOG', 'NEG', "
        "'REAL', 'SIGN'}")
    .Attr("T: {bfloat16, half, float, double}")
    .SetShapeFn(shape_inference::UnchangedShape);

// TODO(b/178234771): change shape function.
REGISTER_OP("RiscWhile")
    .Input("input: T")
    .Output("output: T")
    .Attr("T: list(type) >= 0")
    .Attr("cond: func")
    .Attr("body: func")
    .Attr("output_shapes: list(shape) = []")
    .Attr("parallel_iterations: int = 10")
    .SetIsStateful()
    .SetShapeFn(shape_inference::UnknownShape);

}  // namespace tensorflow

File: tensorflow/python/ops/risc/risc_ops.py
# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""RISC Operations."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function


from tensorflow.python.ops import gen_risc_ops


def risc_abs(x, name='RISC_ABS'):
  return gen_risc_ops.risc_abs(x, name=name)


def risc_add(
    input_lhs,
    input_rhs,
    name='RISC_ADD'):
  return gen_risc_ops.risc_add(input_lhs, input_rhs, name=name)


def risc_binary_arithmetic(x, y, op_type, name='RISC_BinaryArithmetic'):
  return gen_risc_ops.risc_binary_arithmetic(x, y, op_type=op_type, name=name)


def risc_binary_comparison(x, y, op_type, name='RISC_BinaryComparison'):
  return gen_risc_ops.risc_binary_comparison(x, y, op_type=op_type, name=name)


def risc_bitcast(x, dtype, name='RISC_BITCAST'):
  return gen_risc_ops.risc_bitcast(x, dtype, name=name)


def risc_broadcast(x, shape, name='RISC_BROADCAST'):
  return gen_risc_ops.risc_broadcast(x, shape, name=name)


def risc_cast(x, dtype, name='RISC_CAST'):
  return gen_risc_ops.risc_cast(x, dtype, name=name)


def risc_ceil(x, name='RISC_CEIL'):
  return gen_risc_ops.risc_ceil(x, name=name)


def risc_cos(x, name='RISC_COS'):
  return gen_risc_ops.risc_cos(x, name=name)


def risc_cholesky(x, name='RISC_CHOLESKY'):
  return gen_risc_ops.risc_cholesky(x, name=name)


def risc_concat(x, axis, name='RISC_CONCAT'):
  return gen_risc_ops.risc_concat(x, axis, name=name)


def risc_condition(pred,
                   input_true,
                   input_false,
                   func_true,
                   func_false,
                   name='RISC_CONDITION'):
  return gen_risc_ops.risc_condition(
      pred,
      input_true,
      input_false,
      func_true=func_true,
      func_false=func_false,
      name=name)


def risc_conv(x,
              kernel,
              strides,
              data_format='NHWC',
              dilations=None,
              name='RISC_CONV'):
  return gen_risc_ops.risc_conv(
      x,
      kernel,
      strides,
      data_format=data_format,
      dilations=dilations,
      name=name)


def risc_div(input_lhs, input_rhs, name='RISC_DIV'):
  return gen_risc_ops.risc_div(input_lhs, input_rhs, name=name)


def risc_dot(input_lhs,
             input_rhs,
             transpose_a=False,
             transpose_b=False,
             name='RISC_DOT'):
  return gen_risc_ops.risc_dot(
      input_lhs,
      input_rhs,
      transpose_a=transpose_a,
      transpose_b=transpose_b,
      name=name)


def risc_exp(x, name='RISC_EXP'):
  return gen_risc_ops.risc_exp(x, name=name)


def risc_fft(x, name='RISC_FFT'):
  return gen_risc_ops.risc_fft(x, name=name)


def risc_floor(x, name='RISC_FLOOR'):
  return gen_risc_ops.risc_floor(x, name=name)


def risc_gather(params,
                indices,
                validate_indices=None,
                axis=None,
                batch_dims=0,
                name='RISC_GATHER'):
  return gen_risc_ops.risc_gather(
      params,
      indices,
      validate_indices=validate_indices,
      name=name,
      axis=axis,
      batch_dims=batch_dims)


def risc_imag(x, name='RISC_IMAG'):
  return gen_risc_ops.risc_imag(x, name=name)


def risc_is_finite(x, name='RISC_IS_FINITE'):
  return gen_risc_ops.risc_is_finite(x, name=name)


def risc_log(x, name='RISC_LOG'):
  return gen_risc_ops.risc_log(x, name=name)


def risc_logical_and(a, b, name='RISC_LOGICAL_AND'):
  return gen_risc_ops.risc_logical_and(a, b, name=name)


def risc_logical_not(a, b, name='RISC_LOGICAL_NOT'):
  return gen_risc_ops.risc_logical_not(a, b, name=name)


def risc_logical_or(a, b, name='RISC_LOGICAL_OR'):
  return gen_risc_ops.risc_logical_or(a, b, name=name)


def risc_max(input_lhs, input_rhs, name='RISC_MAX'):
  return gen_risc_ops.risc_max(input_lhs, input_rhs, name=name)


def risc_min(input_lhs, input_rhs, name='RISC_MIN'):
  return gen_risc_ops.risc_min(input_lhs, input_rhs, name=name)


def risc_mul(input_lhs, input_rhs, name='RISC_MUL'):
  return gen_risc_ops.risc_mul(input_lhs, input_rhs, name=name)


def risc_neg(x, name='RISC_NEG'):
  return gen_risc_ops.risc_neg(x, name=name)


def risc_pad(x, padding, constant_values, name='RISC_PAD'):
  return gen_risc_ops.risc_pad(x, padding, constant_values, name=name)


def risc_pool(x, ksize, strides, pooling_type='MAX', name='RISC_POOL'):
  return gen_risc_ops.risc_pool(
      x, ksize, strides, pooling_type=pooling_type, name=name)


def risc_pow(input_lhs, input_rhs, name='RISC_POW'):
  return gen_risc_ops.risc_pow(input_lhs, input_rhs, name=name)


def risc_random_uniform(shape, seed, name='RISC_RANDOM_UNIFORM'):
  return gen_risc_ops.risc_random_uniform(shape, seed, name=name)


def risc_real(x, name='RISC_REAL'):
  return gen_risc_ops.risc_real(x, name=name)


def risc_reduce(x, axis, reduce_type, name='RISC_REDUCE'):
  return gen_risc_ops.risc_reduce(x, axis, reduce_type=reduce_type, name=name)


def risc_rem(x, name='RISC_REM'):
  return gen_risc_ops.risc_rem(x, name=name)


def risc_reshape(x, shape, name='RISC_RESHAPE'):
  return gen_risc_ops.risc_reshape(x, shape, name=name)


def risc_reverse(x, axis, name='RISC_REVERSE'):
  return gen_risc_ops.risc_reverse(x, axis, name=name)


def risc_scatter(indices, updates, shape, name='RISC_SCATTER'):
  return gen_risc_ops.risc_scatter(indices, updates, shape, name=name)


def risc_shape(x, name='RISC_SHAPE'):
  return gen_risc_ops.risc_shape(x, name=name)


def risc_sign(x, name='RISC_SIGN'):
  return gen_risc_ops.risc_sign(x, name=name)


def risc_slice(x, begin, size, name='RISC_SLICE'):
  return gen_risc_ops.risc_slice(x, begin, size, name=name)


def risc_sub(input_lhs, input_rhs, name='RISC_SUB'):
  return gen_risc_ops.risc_sub(input_lhs, input_rhs, name=name)


def risc_sort(x, axis, direction='ASCENDING', name='RISC_SORT'):
  return gen_risc_ops.risc_sort(x, axis, direction=direction, name=name)


def risc_squeeze(x, axis=None, name='RISC_SQUEEZE'):
  return gen_risc_ops.risc_squeeze(x, axis, name=name)


def risc_transpose(x, perm=None, name='RISC_TRANSPOSE'):
  return gen_risc_ops.risc_transpose(x, perm, name=name)


def risc_triangular_solve(matrix,
                          rhs,
                          lower=True,
                          adjoint=False,
                          name='RISC_TRIANGULAR_SOLVE'):
  return gen_risc_ops.risc_triangular_solve(
      matrix, rhs, lower=lower, adjoint=adjoint, name=name)


def risc_unary(x, op_type='ABL', name='RISC_UNARY'):
  return gen_risc_ops.risc_unary(x, op_type=op_type, name=name)


def risc_while(cond,
               body,
               loop_vars,
               shape_invariants=None,
               parallel_iterations=10,
               back_prop=True,
               swap_memory=False,
               maximum_iterations=None,
               name='RISC_WHILE'):
  return gen_risc_ops.risc_while(
      cond=cond,
      body=body,
      loop_vars=loop_vars,
      shape_invariants=shape_invariants,
      parallel_iterations=parallel_iterations,
      back_prop=back_prop,
      swap_memory=swap_memory,
      name=name,
      maximum_iterations=maximum_iterations,
      return_same_structure=True)
